### ISSUE TITLE
RFC: Remove features and API from TLA+ tools

### ISSUE THREAD

**Issue by lemmy** [tlaplus/tlaplus#718]
Candidates for removal (number in brackets shows supporters of the feature):

### Feature
- Attach files of all modules in TLC's result email
- DFID (depth-first iterative deepening) model-checking
- Distributed TLC & Cloud TLC
- Checkpointing & recovery of the state queue, fingerprint set, and the trace (`-checkpoint` & `-recover` parameter)
- `-DTLA-Library` Java system property
- [1] Resuming reachability checking after a counter-example has been found (`-continue`  parameter)
- Overriding of built-in standard modules with custom ones
- Loading Java module overrides from class files in spec or `TLA-Library`
- ~SANY#XMLExport~
- Toolbox (in favor of [VSCode extension](https://github.com/tlaplus/vscode-tlaplus))

### API
- `tla2sany.semantic.FrontEnd.getToolId()`
- `tla2sany.modanalyzer.ParseUnit.writeParseTreeToFile(boolean, Errors)` (implemented by Jean-Charles Gregoire)
- `tla2sany.semantic.ModuleNode.isStandard()`, `util.FilenameToStream.TLAFile.isLibraryModule()`, `tla2sany.modanalyzer.ParseUnit.isLibraryModule()`

Related:
- #688 #378
- `tla2sany.modanalyzer.SpecObj.validateParseUnit(ParseUnit)` (remove `FileInputStream` from API) (`ParseUnit#NamedInputStream` is closed when `validateParseUnit` is called)

**Comment by eras** [comment]
So these listed features are candidates for removal?

I suppose it's a bit sad to see all these features go, granted I had never used any of them but for two of them: the Toolbox (nowadays I use Emacs mostly), and the `-continue` switch. I do wonder if newcomers would find the VSCode Extension easier to approach, but I've never given it a try.

The `-continue` switch I've used to create a bunch of "example cases" from a state machine. I built support for handling that kind of output to my newly released tool for generating sequence diagrams out of state traces: https://github.com/eras/tlsd . Sometimes you might want to see more examples than just the first one (so basically random) and maybe gain some insight that way.

Maybe Distributed TLC is not that useful nowadays when you can get big machines affordably from e.g. Amazon for short runs, and if I recall correctly, DTLC was not able to check for liveness properties anyway.

**Comment by zwergziege** [comment]
I tend to agree with eras. Would the removal of those features really do much in terms of maintainability / code quality?
Concerning `-checkpoint` and distributed TLC: I have never used them, but can understand why someone would want to use them. Especially checkpoints.
Concerning Java module overrides and library path: Will there be any alternatives? If not, wouldn't that break the community modules, or would the community modules be kept as the only library file? Also, the proof lib (as it is now) depends on library path capabilities. So if you want to model check something you're also doing proofs on, you'll need that. Oc it's not too hard to just copy the files in your main folder. 
Concerning the toolbox: Is there any plan to incorporate TLAPM into the VSCode extension? On the other hand, I do like the idea of not having several tools that do the same thing.

**Comment by andrey-kuprianov** [comment]
Similar to @eras, we do quite heavily rely on `-continue` flag in the [modelator](https://github.com/informalsystems/modelator) model-based testing tool to generate multiple test cases from the same test assertion.

@lemmy would it be possible to maintain this feature, or provide an alternative one with similar semantics?

**Comment by danwt** [comment]
If `-DTLA-Library Java system property` I would still like custom java operators to remain a feature!

**Comment by lemmy** [comment]
> Similar to @eras, we do quite heavily rely on `-continue` flag in the [modelator](https://github.com/informalsystems/modelator) model-based testing tool to generate multiple test cases from the same test assertion.
> 
> @lemmy would it be possible to maintain this feature, or provide an alternative one with similar semantics?

@andrey-kuprianov Has it ever been an issue that `-continue` does not work for violations that can only be found by checking the behavior graph instead of the state graph (simpler safety properties)?

> If `-DTLA-Library Java system property` is removed, I would still like custom java operators to remain a feature, somehow!

@danwt I'm not proposing the removal of custom Java operators, but only one (the old) mechanism to load them.  Also, the newer mechanism is more robust and feature-rich (@TLAPlusOperator, @Evaluation, ...).

**Comment by arssher** [comment]
I quite wanted to use `-recover` (checkpointing) for big models which can run for weeks in case e.g. server restarts, or to move to another server without losing progress. However, at least on 1.8.0 they don't seem to be really usable, most often it fails with different errors.

**Comment by goyalyashpal** [comment]
> * Toolbox (in favor of [VSCode extension](https://github.com/tlaplus/vscode-tlaplus))
> \- _@lemmy at https://github.com/tlaplus/tlaplus/issues/718#issue-1155484851_

i use toolbox and am liking it - breath of fresh air.

* flexible arrangement of panes/views 
this helps for example for viewing specs, model, errors, etc simultaneously in whatever custom layout we want.

* fully functional / fully developed view for model (configuration, stats, etc.) and error trace (of actions, & values with go to line).

  i've seen _very few_ if any vscode extensions which can match that kind of rich and mature gui in vscode's main editor area.  most vscode extensions have to fiddle with its restricted side bars.

  also see: https://github.com/microsoft/vscode/issues/198611

* dedicated / tailor made help links to book, tutorial etc.
in tla+ toolbox, i can be sure that any link etc i click on will most likely help me in some regards with tla/tlc etc.

  this sureity won't be possible with vscode.

* vscode follows a closed tightly choked development, and it's aim is NOT to cater to community at all.
it's a commercial product with commercial goal.  so, unless tlaplus has some internal dealings with vscode team, it would be much much hard to get any consideration for flexibility / helping features in there.

* if tla+ toolbox were to get abandoned (_"Remove ... toolbox (in favour of ...)"_), it would be a cognitive hinderance to use an unmaintained / standby-mode application to follow leslie lamport's awesome "tla+ video course".  you don't get _video_ mooc course from famous computer scientists every day.

**Comment by konnov** [comment]
I just found that `tla2tools.jar` declares `mailapi` as a dependency, which is kind of unusual for a model checker. Codex tells me that this is needed by a legacy e-mailing feature. Just flagging it here, maybe it's a good idea to remove mailapi completely.
